src/code42cli/cmds/devices.py

```diff
@@ -0,0 +1,396 @@
+    )
+    device["lastCompletedBackup"] = (
+        max(
+            [
+                backupDestination["lastCompletedBackup"]
+                for backupDestination in device["backupUsage"]
+            ]
+        )
+        if len(device["backupUsage"]) > 0
+        else None
+    )
+    return device
+
+
+@devices.command(name="list", help="Get information about many devices")
+@click.option(
+    "--active",
+    required=False,
+    type=bool,
+    is_flag=True,
+    default=None,
+    help="Include to get only active or deactivated devices. Defaults to getting all devices.",
+)
+@click.option(
+    "--days-since-last-connected",
+    required=False,
+    type=int,
+    help="Return only devices that have not connected in the number of days specified.",
+)
+@click.option(
+    "--org-uid",
+    required=False,
+    type=str,
+    default=None,
+    help="""Optionally provide to limit devices to only the ones in the org you specify.
+    Note that child orgs will be included.""",
+)
+@click.option(
+    "--drop-most-recent",
+    required=False,
+    type=int,
+    help="""Will drop the X most recently connected devices for each user from the
+    result list where X is the number you provide as this argument. Can be used to
+    avoid passing the most recently connected device for a user to the deactivate command.""",
+)
+@click.option(
+    "--include-backup-usage",
+    required=False,
+    type=bool,
+    default=False,
+    is_flag=True,
+    help="""Include to return backup usage information for each device
+    (may significantly lengthen the size of the return).""",
+)
+@click.option(
+    "--include-usernames",
+    required=False,
+    type=bool,
+    default=False,
+    is_flag=True,
+    help="Include to add the username associated with a device to the output.",
+)
+@click.option(
+    "--include-settings",
+    required=False,
+    type=bool,
+    default=False,
+    is_flag=True,
+    help="Include to include device settings in output.",
+)
+@sdk_options()
+def list(
+    state,
+    active,
+    days_since_last_connected,
+    drop_most_recent,
+    org_uid,
+    include_backup_usage,
+    include_usernames,
+    include_settings,
+):
+    """Outputs a list of all devices."""
+    devices_dataframe = _get_device_dataframe(
+        state.sdk, active, org_uid, include_backup_usage
+    )
+    if drop_most_recent:
+        devices_dataframe = _drop_n_devices_per_user(
+            devices_dataframe, drop_most_recent
+        )
+    if days_since_last_connected:
+        devices_dataframe = _drop_devices_which_have_not_connected_in_some_number_of_days(
+            devices_dataframe, days_since_last_connected
+        )
+    if include_settings:
+        devices_dataframe = _add_settings_to_dataframe(state.sdk, devices_dataframe)
+    if include_usernames:
+        devices_dataframe = _add_usernames_to_device_dataframe(
+            state.sdk, devices_dataframe
+        )
+    click.echo_via_pager(devices_dataframe.to_csv())
```

Comment:
Our other `list` commands have a convention of multiple format options (with the default being a human-readable table). 

But.. This has me thinking. DataFrames have a really convenient `.to_string()` method that basically does the equivalent to our hand-rolled table output formatter. 

We could adopt the use of dataframes for all of our commands that need to output tabular results and simplify our codebase a ton (in addition to enabling other features across the board).

Reply:
That sounds great! Should we get this PR through first and then refactor? We can do it as a quick follow to this.

Reply:
I've added a formatter for dataframes, and changed the `list` method to use it. There isn't much there but it can use the same format option as the other methods that accept different formats.